### PR TITLE
add extra test for the hashindex

### PR DESCRIPTION
**Pull request by rciorba**
Insert a few keys, delete some of them, check we still have
the values we expect, check the deleted ones aren't there.

### CODE REVIEW

File: src/borg/testsuite/hashindex.py [hunk 1]
```diff
@@ -125,6 +125,31 @@ def test_chunkindex_summarize(self):
         assert unique_chunks == 3
 
 
+class HashIndexExtraTestCase(BaseTestCase):
+    """These tests are separate because they should not become part of the selftest
+    """
+    def test_chunk_indexer(self):
+        # see _hashindex.c hash_sizes, we want to be close to the max fill rate
+        # because interesting errors happen there
+        max_key = int(65537 * 0.75) - 10
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/hashindex.py]: import/use MAX_LOAD_FACTOR (or so)?

File: src/borg/testsuite/hashindex.py [hunk 2]
```diff
@@ -125,6 +125,31 @@ def test_chunkindex_summarize(self):
         assert unique_chunks == 3
 
 
+class HashIndexExtraTestCase(BaseTestCase):
+    """These tests are separate because they should not become part of the selftest
+    """
+    def test_chunk_indexer(self):
+        # see _hashindex.c hash_sizes, we want to be close to the max fill rate
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/hashindex.py]: s/max fill rate/max. load/

File: src/borg/testsuite/hashindex.py [hunk 3]
```diff
@@ -125,6 +125,31 @@ def test_chunkindex_summarize(self):
...
+        # because interesting errors happen there
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/hashindex.py]: nitpick: add dot at end of sentence. also in 129.

File: src/borg/testsuite/hashindex.py [hunk 4]
```diff
@@ -125,6 +125,31 @@ def test_chunkindex_summarize(self):
...
+        max_key = int(65537 * 0.75) - 10
+        index = ChunkIndex(max_key)
+        deleted_keys = [
+            hashlib.sha256(H(k)).digest()
+            for k in range(-1, -(max_key//3), -1)]
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/hashindex.py]: if we are still < 120, put it in 1 line? no 80 chars pep8 limit.

File: src/borg/testsuite/hashindex.py [hunk 5]
```diff
@@ -125,6 +125,31 @@ def test_chunkindex_summarize(self):
...
+        for i, key in enumerate(deleted_keys):
+            index[key] = (i, i, i)
+
+        for key in deleted_keys:
+            del index[key]
+        for i, key in enumerate(keys):
+            assert index[key] == (i, i, i)
+        for key in deleted_keys:
+            assert index.get(key) is None
+
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/hashindex.py]: now delete all keys also and make sure all are gone?

File: src/borg/testsuite/hashindex.py (hunk 1 above)
Comment (rciorba) [inline src/borg/testsuite/hashindex.py]: How can I do that? As far as i know it's a C macro, and there is (still) no way to reach it.

File: src/borg/testsuite/hashindex.py (hunk 5 above)
Comment (rciorba) [inline src/borg/testsuite/hashindex.py]: I don't understand what you mean. 

Do you mean add a comment explaining what's going on? I assumed the code is simple and self explanatory.

File: src/borg/testsuite/hashindex.py (hunk 1 above)
Comment (ThomasWaldmann) [inline src/borg/testsuite/hashindex.py]: iirc we made it available via cython and already use it elsewhere.

File: src/borg/testsuite/hashindex.py (hunk 5 above)
Comment (ThomasWaldmann) [inline src/borg/testsuite/hashindex.py]: you are only deleting `deleted_keys` to test insert / delete behaviour with full hashtable / growing hashtable (/ shrinking hashtable).

one could also delete keys from `keys` again, just to have some more tests.

**Comment by ThomasWaldmann** [comment]
@rciorba can we finish this soon to get it off the table? seems like only some simple changes are still todo.

**Comment by ThomasWaldmann** [comment]
thanks for updating, looks good. \o/

I'll wait for travis and merge then.

**Pull request merged**